File: lib/adventofcode.ex
defmodule Adventofcode do
  def y2020d1 do
    {:ok, input } = File.read("lib/adventofcode/2020/inputs/1.txt")

    input
      |> String.split("\n", trim: true)
      |> Enum.map(&String.to_integer/1)
     # |> permutations_of()
      |> Enum.find(fn x -> Enum.sum(x) == 2994 end)
      |> Enum.reduce(&Kernel.*/2)

    # RUBY
        # input.combination(2) do |comb|
        #   int_arr = comb.map(&:to_i)
        #   return int_arr[0] * int_arr[1] if int_arr.sum == 2020
        # end
  end

  def y2017d1p1 do
#    {:ok, input } = File.read("lib/adventofcode/2007/inputs/1.txt")
    input = [1,1,2,2]

    input
      |> String.split("\n", trim: true)
      |> Enum.map(&String.to_integer/1)
  end
end

File: lib/adventofcode.rb
# require "adventofcode/version"
require 'pry'

module Adventofcode
  def self.part_one(year, day, input = nil)
    require File.join(File.absolute_path(File.dirname(__FILE__)), 'adventofcode', year, day)
    Object.const_get("Adventofcode::Year#{year}::Day#{day}").new(year, day, input).part_one
  end

  def self.part_two(year, day, input = nil)
    require File.join(File.absolute_path(File.dirname(__FILE__)), 'adventofcode', year, day)
    Object.const_get("Adventofcode::Year#{year}::Day#{day}").new(year, day, input).part_two
  end

  class Error < StandardError; end

  class Day
    attr_reader :year, :day, :input

    def initialize(year, day, input)
      @year  = year
      @day   = day
      @input = input || default_input
    end

    def default_input
      raw_input.split("\n").map(&:to_i)
    end

    def raw_input
      File.read(File.join(File.absolute_path(File.dirname(__FILE__)), 'adventofcode', year, 'inputs', "#{day}.txt"))
    end
  end
end
